src/gps/nmea_parser.cc

```diff
@@ -0,0 +1,137 @@
+// BoB robotics includes
+#include "common/macros.h"
+#include "gps/nmea_parser.h"
+
+namespace BoBRobotics {
+namespace GPS {
+
+void
+parseTimeField(const std::string &field, std::tm &time)
+{
+    time.tm_hour = stoi(field.substr(0, 2));
+    time.tm_min = stoi(field.substr(2, 2));
+    time.tm_sec = stoi(field.substr(4, 2));
+
+    // Indicates that information about DST is not available
+    time.tm_isdst = -1;
+}
+
+bool
+NMEAParser::parseDateTime(const std::string &line, std::tm &time)
+{
+    if (!parse(line, "$GNZDA", 6)) {
+        return false;
+    }
+
+    try {
+        parseTimeField(m_Fields[0], time);
+        time.tm_mday = stoi(m_Fields[1]);
+        time.tm_mon = stoi(m_Fields[2]) - 1;
+        time.tm_year = stoi(m_Fields[3]) - 1900;
+
+        // This field is available in GNU/BSD but is not strict ISO C
+        int hoursOff = m_Fields[4].empty() ? 0 : stoi(m_Fields[4]);
+        int minutesOff = m_Fields[5].size() <= 3 ? 0 : stoi(m_Fields[5].substr(0, 2));
+        time.tm_gmtoff = 3600 * hoursOff + 60 * minutesOff;
+    } catch (std::invalid_argument &e) {
```

Comment:
I've never actually done it before but I _think_ the correct thing to do here would be to 
```c++
catch (const std::invalid_argument &e) {
    std::throw_with_nested(NMEAError());
}
catch (const std::out_of_range &e) {
    std::throw_with_nested(NMEAError());
}
```
Then wherever you catch it:
```c++
catch(const NMEAError &e) {
   LOGW << "NMEA parsing error: " << m_Line;
    try {
        std::rethrow_if_nested(e);
    } catch(const std::exception& e) {
        LOGW << "NMEA parsing error internal:" <<  e.what();
    }
}
```
That way you get to keep the single exception type to throw but you aren't throwing away the internals which might be helpful for your debugger.

Reply:
Oh cool! I didn't know this was a thing

Reply:
hah me neither - C++ exception-handling seems an infinitely deep mire of complexity